packages/core/src/modules/oob/OutOfBandMessage.ts

```diff
@@ -0,0 +1,90 @@
+import type { UnpackedMessage } from '../../types'
+import type { DidCommService } from '../connections/models/did/service/DidCommService'
+
+import { Expose, Type } from 'class-transformer'
+import { Equals, IsArray, IsInstance, ValidateNested } from 'class-validator'
+
+import { AgentMessage } from '../../agent/AgentMessage'
+import { Attachment, AttachmentData } from '../../decorators/attachment/Attachment'
+import { JsonEncoder } from '../../utils/JsonEncoder'
+
+const VERSION = '1.1'
+
+export interface OutOfBandMessageOptions {
+  id?: string
+  label?: string
+  goalCode?: string
+  goal?: string
+}
+
+export class OutOfBandMessage extends AgentMessage {
+  public constructor(options: OutOfBandMessageOptions) {
+    super()
+    if (options) {
+      this.id = options.id ?? this.generateId()
+      this.label = options.label
+      this.goalCode = options.goalCode
+      this.goal = options.goal
+    }
+  }
+
+  public addHandshakeProtocol(protocol: string) {
+    if (!this.handshakeProtocols) this.handshakeProtocols = []
+    this.handshakeProtocols.push(protocol)
+  }
+
+  public addRequest(message: AgentMessage) {
+    if (!this.requests) this.requests = []
+    const requestAttachment = new Attachment({
+      id: 'request-0',
+      mimeType: 'application/json',
+      data: new AttachmentData({
+        base64: JsonEncoder.toBase64(message.toJSON()),
+      }),
+    })
+    this.requests.push(requestAttachment)
+  }
+
+  public getRequests(): UnpackedMessage[] | undefined {
+    return this.requests?.map((request) => {
+      if (request.data.base64) {
+        return JsonEncoder.fromBase64(request.data.base64)
```

Comment:
We should add something to handle attachments in a generic way (ACA-Py has it). Data could be base64 or json (or others but the first two are most important)

```ts
const data = request.data.get() // could be either json or base64, but it doesn't matter
```

Reply:
That could be good, at least for `base64` and `json`. However, if I look at [Aries RFC 0017: Attachments](https://github.com/hyperledger/aries-rfcs/blob/main/concepts/0017-attachments/README.md) it can also contain `links`. Should we call some HTTP request in that case or just return the link?

Reply:
This perhaps should not be implemented as a part of this PR, right?

Reply:
I added also if statement for `json` and created issue #548